patches/tock/09-add-vendor-hid-usb-interface.patch

```diff
@@ -0,0 +1,331 @@
+ 
+@@ -82,6 +120,10 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+         product_id: u16,
+         strings: &'static [&'static str],
+     ) -> Self {
++        if cfg!(feature = "vendor_hid") {
++            debug!("vendor_hid enabled.");
++        }
++
+         let interfaces: &mut [InterfaceDescriptor] = &mut [
+             // Interface declared in the FIDO2 specification, section 8.1.8.1
+             InterfaceDescriptor {
+@@ -90,9 +132,19 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+                 interface_protocol: 0x00,
+                 ..InterfaceDescriptor::default()
+             },
++            // Vendor HID interface.
++            #[cfg(feature = "vendor_hid")]
++            InterfaceDescriptor {
++                interface_number: 1,
++                interface_class: 0x03, // HID
++                interface_subclass: 0x00,
++                interface_protocol: 0x00,
++                ..InterfaceDescriptor::default()
++            },
+         ];
+ 
+         let endpoints: &[&[EndpointDescriptor]] = &[&[
++            // 2 Endpoints for FIDO
+             EndpointDescriptor {
+                 endpoint_address: EndpointAddress::new_const(
+                     ENDPOINT_NUM,
+@@ -110,8 +162,30 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+                 transfer_type: TransferType::Interrupt,
+                 max_packet_size: 64,
+                 interval: 5,
+-            },
+-        ]];
++            },],
++            // 2 Endpoints for FIDO
++            #[cfg(feature = "vendor_hid")]
++            &[
++                EndpointDescriptor {
++                    endpoint_address: EndpointAddress::new_const(
++                        ENDPOINT_NUM + 1,
++                        TransferDirection::HostToDevice,
++                    ),
++                    transfer_type: TransferType::Interrupt,
++                    max_packet_size: 64,
++                    interval: 5,
++                },
++                EndpointDescriptor {
++                    endpoint_address: EndpointAddress::new_const(
++                        ENDPOINT_NUM + 1,
++                        TransferDirection::DeviceToHost,
++                    ),
++                    transfer_type: TransferType::Interrupt,
++                    max_packet_size: 64,
++                    interval: 5,
++                },
++            ],
++        ];
+ 
+         let (device_descriptor_buffer, other_descriptor_buffer) =
+             descriptors::create_descriptor_buffers(
+@@ -130,17 +204,28 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+                 },
+                 interfaces,
+                 endpoints,
+-                Some(&HID),
++                Some(&[&[
++                    &HID,
++                    #[cfg(feature = "vendor_hid")]
++                    &HID,
++                ]]),
```

Comment:
I don't understand why we have 2 layers of descriptors, i.e. `&[&[&X, &X]]` instead of `&[&X, &X]`? In particular you always only look at the first element of the outer layer, which suggests the outer layer is a singleton and could be omitted. Or do you foresee usage in the future? Also, it's probably worth documenting the meaning of those layers. I'll add a comment at the related locations with the marker LAYERS.

Reply:
This approach was used to match the approach used for the endpoints:
https://github.com/tock/tock/blob/master/capsules/src/usb/descriptors.rs#L423

I'm happy to take on other  suggestions.

Reply:
:facepalm: 
The endpoints are nested because there are multiple endpoints per interface.
For the descriptors, only one is needed per interfaces, so the extra layer can be removed.


I'll fix that up and update the PR.

Reply:
OK, new commits have removed this. PTAL.